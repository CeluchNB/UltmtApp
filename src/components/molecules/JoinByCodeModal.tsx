import BaseModal from '../atoms/BaseModal'
import PrimaryButton from '../atoms/PrimaryButton'
import React from 'react'
import UserInput from '../atoms/UserInput'
import { getFormFieldRules } from '../../utils/form-utils'
import { useTheme } from '../../hooks'
import validator from 'validator'
import { Controller, useForm } from 'react-hook-form'
import { StyleSheet, Text, View } from 'react-native'

interface JoinByCodeModalProps {
    visible: boolean
    loading: boolean
    error: string
    onSubmit: (data: { code: string }) => void
    onClose: () => void
}

const JoinByCodeModal: React.FC<JoinByCodeModalProps> = ({
    visible,
    loading,
    error,
    onSubmit,
    onClose,
}) => {
    const {
        theme: { colors, size },
    } = useTheme()
    const {
        control,
        handleSubmit,
        reset,
        formState: { errors },
    } = useForm({ defaultValues: { code: '' } })

    const styles = StyleSheet.create({
        header: {
            fontSize: size.fontThirty,
            color: colors.textPrimary,
        },
        inputContainer: {
            display: 'flex',
            flexDirection: 'row',
            marginTop: 10,
        },
        errorText: {
            fontSize: size.fontFifteen,
            color: colors.error,
        },
        button: {
            marginTop: 10,
        },
    })

    return (
        <BaseModal visible={visible} onClose={onClose}>
            <Text style={styles.header}>Join With Code</Text>
            <View style={styles.inputContainer}>
                <Controller
                    name="code"
                    control={control}
                    rules={getFormFieldRules('Code', true, 6, 6, [
                        {
                            test: (v: string) => {
                                return validator.isNumeric(v)
                            },
                            message: 'Code must be numeric',
                        },
                    ])}
                    render={({ field: { onChange, value } }) => {
                        return (
                            <UserInput
                                placeholder="6 Digit Code"
                                onChangeText={onChange}
                                value={value}
                            />
                        )
                    }}
                />
            </View>
            {errors.code && (
                <Text style={styles.errorText}>{errors.code.message}</Text>
            )}
            {error.length > 0 && <Text style={styles.errorText}>{error}</Text>}
            <PrimaryButton
                style={styles.button}
                text="join"
                onPress={() => {
                    handleSubmit(onSubmit)()
                    reset()
                }}
                loading={loading}
            />
        </BaseModal>
    )
}

export default JoinByCodeModal
