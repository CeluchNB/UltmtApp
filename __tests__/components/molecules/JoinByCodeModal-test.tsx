import JoinByCodeModal from '../../../src/components/molecules/JoinByCodeModal'
import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')
jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter')

describe('JoinByCodeModal', () => {
    it('should match snapshot', () => {
        const snapshot = render(
            <JoinByCodeModal
                visible={true}
                onClose={jest.fn()}
                loading={false}
            />,
        )

        expect(snapshot.getByPlaceholderText('6 Digit Code')).toBeTruthy()

        expect(snapshot).toMatchSnapshot()
    })

    it('should call on close', async () => {
        const onClose = jest.fn()
        const { getByText, getByPlaceholderText } = render(
            <JoinByCodeModal
                visible={true}
                onClose={onClose}
                loading={false}
            />,
        )

        const input = getByPlaceholderText('6 Digit Code')
        fireEvent.changeText(input, '123456')

        const button = getByText('join')
        fireEvent.press(button)

        await waitFor(() => {
            expect(onClose).toBeCalledWith({ code: '123456' }, undefined)
        })
    })

    it('should handle error', async () => {
        const onClose = jest.fn()
        const { getByText } = render(
            <JoinByCodeModal
                visible={true}
                onClose={onClose}
                loading={false}
            />,
        )

        const button = getByText('join')
        fireEvent.press(button)

        await waitFor(() => {
            expect(getByText('Code is required.')).toBeTruthy()
        })
        expect(onClose).not.toHaveBeenCalled()
    })
})
