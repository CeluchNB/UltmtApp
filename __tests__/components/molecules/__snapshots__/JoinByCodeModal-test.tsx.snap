// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinByCodeModal should match snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  testID="base-modal"
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#30303050",
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#1a1a1a",
          "borderColor": "#ffd600",
          "borderRadius": 10,
          "color": "#ffd600",
          "elevation": 5,
          "flex": 1,
          "margin": 20,
          "padding": 25,
          "shadowColor": "#ffd600",
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffd600",
            "fontSize": 61.64383561643835,
          }
        }
      >
        Join With Code
      </Text>
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "marginTop": 10,
          }
        }
      >
        <View
          style={{}}
        >
          <View
            pointerEvents="none"
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                {
                  "backgroundColor": "#303030",
                  "borderColor": "#7AC6FF",
                  "borderRadius": 4,
                  "borderWidth": 1,
                },
                undefined,
              ]
            }
            testID="text-input-outline"
          />
          <View>
            <View
              style={
                [
                  {
                    "paddingBottom": 0,
                  },
                  {
                    "minHeight": 56,
                    "paddingTop": 8,
                  },
                ]
              }
            >
              <TextInput
                accessibilityLabel="6 Digit Code"
                autoCapitalize="none"
                editable={true}
                keyboardType="default"
                maxFontSizeMultiplier={1.5}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="6 Digit Code"
                placeholderTextColor="#ffffff"
                selectionColor="#ffd600"
                style={
                  [
                    {
                      "margin": 0,
                      "zIndex": 1,
                    },
                    {
                      "height": 48,
                    },
                    {
                      "paddingBottom": 0,
                      "paddingTop": 0,
                    },
                    {
                      "color": "#ffd600",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": undefined,
                      "letterSpacing": 0.15,
                      "lineHeight": undefined,
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                      "textAlignVertical": "center",
                    },
                    false,
                    [
                      {},
                    ],
                    undefined,
                  ]
                }
                testID="text-input-outlined"
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          {
            "alignSelf": undefined,
            "bottom": undefined,
            "end": undefined,
            "left": undefined,
            "position": undefined,
            "right": undefined,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "start": undefined,
            "top": undefined,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#ffd600",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "marginTop": 10,
                "minWidth": "auto",
              }
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  false,
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "textAlign": "left",
                      },
                      {
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        {
                          "marginHorizontal": 8,
                        },
                        {
                          "textTransform": "uppercase",
                        },
                        {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  join
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
